File: app/routes/reader.py
"""Reader API routes: chapters list, chapter content (with purification), reading progress."""
from flask import Blueprint, request, jsonify
from sqlalchemy.orm import Session

from config import Config
from database import db_session
from models import Novel, Chapter, ReadingProgress, PurificationRule
from services.purifier import purify

reader_bp = Blueprint("reader", __name__)


@reader_bp.route("/novels/<int:novel_id>/chapters", methods=["GET"])
def list_chapters(novel_id: int):
    """Get chapter list (table of contents) for a novel."""
    session: Session = db_session()
    try:
        novel = session.query(Novel).filter_by(
            id=novel_id, user_id=Config.DEFAULT_USER_ID
        ).first()
        if not novel:
            return jsonify({"error": "Novel not found"}), 404

        chapters = (
            session.query(Chapter)
            .filter_by(novel_id=novel_id)
            .order_by(Chapter.chapter_index)
            .all()
        )

        # Get enabled purification rules
        active_rules = (
            session.query(PurificationRule)
            .filter_by(user_id=Config.DEFAULT_USER_ID, is_enabled=True)
            .order_by(PurificationRule.order.asc())
            .all()
        )

        result = []
        for ch in chapters:
            title = purify(ch.title, active_rules, scope="title", book_title=novel.title)

            result.append({
                "index": ch.chapter_index,
                "title": title,
                "wordCount": ch.word_count,
            })

        return jsonify(result)
    finally:
        session.close()


@reader_bp.route("/novels/<int:novel_id>/chapters/<int:chapter_index>", methods=["GET"])
def get_chapter(novel_id: int, chapter_index: int):
    """Get a single chapter's content, purified by active rules."""
    session: Session = db_session()
    try:
        novel = session.query(Novel).filter_by(
            id=novel_id, user_id=Config.DEFAULT_USER_ID
        ).first()
        if not novel:
            return jsonify({"error": "Novel not found"}), 404

        chapter = (
            session.query(Chapter)
            .filter_by(novel_id=novel_id, chapter_index=chapter_index)
            .first()
        )
        if not chapter:
            return jsonify({"error": "Chapter not found"}), 404

        # Get enabled purification rules
        active_rules = (
            session.query(PurificationRule)
            .filter_by(user_id=Config.DEFAULT_USER_ID, is_enabled=True)
            .order_by(PurificationRule.order.asc())
            .all()
        )

        title = purify(chapter.title, active_rules, scope="title", book_title=novel.title)
        content = purify(chapter.content, active_rules, scope="content", book_title=novel.title)

        # Get total chapter count for navigation
        total_chapters = (
            session.query(Chapter)
            .filter_by(novel_id=novel_id)
            .count()
        )

        return jsonify({
            "index": chapter.chapter_index,
            "title": title,
            "content": content,
            "wordCount": len(content),
            "totalChapters": total_chapters,
            "hasPrev": chapter.chapter_index > 0,
            "hasNext": chapter.chapter_index < total_chapters - 1,
        })
    finally:
        session.close()


@reader_bp.route("/novels/<int:novel_id>/reading-progress", methods=["GET"])
def get_progress(novel_id: int):
    """Get reading progress for a novel."""
    session: Session = db_session()
    try:
        progress = (
            session.query(ReadingProgress)
            .filter_by(novel_id=novel_id, user_id=Config.DEFAULT_USER_ID)
            .first()
        )
        if not progress:
            return jsonify({
                "chapterIndex": 0,
                "scrollPosition": 0,
                "viewMode": "summary",
            })

        return jsonify({
            "chapterIndex": progress.chapter_index,
            "scrollPosition": progress.scroll_position,
            "viewMode": progress.view_mode,
        })
    finally:
        session.close()


@reader_bp.route("/novels/<int:novel_id>/reading-progress", methods=["PUT"])
def save_progress(novel_id: int):
    """Save reading progress for a novel."""
    data = request.get_json()
    if not data:
        return jsonify({"error": "No data provided"}), 400

    session: Session = db_session()
    try:
        progress = (
            session.query(ReadingProgress)
            .filter_by(novel_id=novel_id, user_id=Config.DEFAULT_USER_ID)
            .first()
        )

        if not progress:
            progress = ReadingProgress(
                user_id=Config.DEFAULT_USER_ID,
                novel_id=novel_id,
            )
            session.add(progress)

        if "chapterIndex" in data:
            progress.chapter_index = data["chapterIndex"]
        if "scrollPosition" in data:
            progress.scroll_position = data["scrollPosition"]
        if "viewMode" in data:
            progress.view_mode = data["viewMode"]

        session.commit()
        return jsonify({"message": "Progress saved"})
    except Exception as e:
        session.rollback()
        return jsonify({"error": str(e)}), 500
    finally:
        session.close()

File: src/api/reader.ts
import { db } from '../services/db';
import type { Chapter as DbChapter } from '../services/db';
import { purify } from '../services/purifier';

export interface Chapter {
  index: number;
  title: string;
  wordCount: number;
}

export interface ChapterContent extends Chapter {
  content: string;
  totalChapters: number;
  hasPrev: boolean;
  hasNext: boolean;
}

export interface ReadingProgress {
  chapterIndex: number;
  scrollPosition: number;
  viewMode: 'summary' | 'original';
  chapterProgress?: number;
  isTwoColumn?: boolean;
}

async function getPurifyRules(): Promise<Array<Record<string, unknown>>> {
  const rules = await db.purificationRules.filter(r => r.isEnabled).sortBy('order');
  return rules.map(r => ({
    name: r.name,
    pattern: r.pattern,
    replacement: r.replacement,
    is_regex: r.isRegex,
    is_enabled: r.isEnabled,
    order: r.order,
    scope_title: r.scopeTitle,
    scope_content: r.scopeContent,
    book_scope: r.bookScope,
    exclude_book_scope: r.excludeBookScope,
  }));
}

async function getNovelTitle(novelId: number): Promise<string> {
  const novel = await db.novels.get(novelId);
  if (!novel) throw new Error('Novel not found');
  return novel.title;
}

export async function loadAndPurifyChapters(novelId: number): Promise<DbChapter[]> {
  const novelTitle = await getNovelTitle(novelId);
  const rawChapters = await db.chapters.where('novelId').equals(novelId).sortBy('chapterIndex');
  const rules = await getPurifyRules();
  if (rules.length === 0) return rawChapters;
  return rawChapters.map(ch => ({
    ...ch,
    title: purify(ch.title, rules, 'title', novelTitle),
    content: purify(ch.content, rules, 'content', novelTitle),
  }));
}

export const readerApi = {
  getChapters: async (novelId: number): Promise<Chapter[]> => {
    const novelTitle = await getNovelTitle(novelId);
    const rawChapters = await db.chapters.where('novelId').equals(novelId).sortBy('chapterIndex');
    const rules = await getPurifyRules();
    if (rules.length === 0) {
      return rawChapters.map(ch => ({ index: ch.chapterIndex, title: ch.title, wordCount: ch.wordCount }));
    }
    return rawChapters.map(ch => ({
      index: ch.chapterIndex,
      title: purify(ch.title, rules, 'title', novelTitle),
      wordCount: ch.wordCount,
    }));
  },

  getChapterContent: async (novelId: number, chapterIndex: number): Promise<ChapterContent> => {
    const novelTitle = await getNovelTitle(novelId);
    const chapter = await db.chapters
      .where('[novelId+chapterIndex]')
      .equals([novelId, chapterIndex])
      .first();
    if (!chapter) throw new Error('Chapter not found');
    const totalChapters = await db.chapters.where('novelId').equals(novelId).count();
    const rules = await getPurifyRules();
    let { title, content } = chapter;
    if (rules.length > 0) {
      title = purify(title, rules, 'title', novelTitle);
      content = purify(content, rules, 'content', novelTitle);
    }
    return {
      index: chapter.chapterIndex,
      title,
      content,
      wordCount: chapter.wordCount,
      totalChapters,
      hasPrev: chapterIndex > 0,
      hasNext: chapterIndex < totalChapters - 1,
    };
  },

  getProgress: async (novelId: number): Promise<ReadingProgress> => {
    const progress = await db.readingProgress.where('novelId').equals(novelId).first();
    if (!progress) {
      return {
        chapterIndex: 0,
        scrollPosition: 0,
        viewMode: 'original',
        chapterProgress: 0,
        isTwoColumn: false,
      };
    }
    return {
      chapterIndex: progress.chapterIndex,
      scrollPosition: progress.scrollPosition,
      viewMode: (progress.viewMode || 'original') as 'summary' | 'original',
      chapterProgress: typeof progress.chapterProgress === 'number' ? progress.chapterProgress : undefined,
      isTwoColumn: typeof progress.isTwoColumn === 'boolean' ? progress.isTwoColumn : undefined,
    };
  },

  saveProgress: async (novelId: number, data: Partial<ReadingProgress>): Promise<{ message: string }> => {
    const existing = await db.readingProgress.where('novelId').equals(novelId).first();
    const now = new Date().toISOString();
    if (existing) {
      await db.readingProgress.update(existing.id, {
        chapterIndex: data.chapterIndex ?? existing.chapterIndex,
        scrollPosition: data.scrollPosition ?? existing.scrollPosition,
        viewMode: data.viewMode ?? existing.viewMode,
        chapterProgress: data.chapterProgress ?? existing.chapterProgress,
        isTwoColumn: data.isTwoColumn ?? existing.isTwoColumn,
        updatedAt: now,
      });
    } else {
      await db.readingProgress.add({
        id: undefined as unknown as number,
        novelId,
        chapterIndex: data.chapterIndex ?? 0,
        scrollPosition: data.scrollPosition ?? 0,
        viewMode: data.viewMode ?? 'original',
        chapterProgress: data.chapterProgress,
        isTwoColumn: data.isTwoColumn,
        updatedAt: now,
      });
    }
    return { message: 'Progress saved' };
  },

  getImageUrl: async (novelId: number, imageKey: string): Promise<string | null> => {
    const image = await db.chapterImages
      .where('[novelId+imageKey]')
      .equals([novelId, imageKey])
      .first();
    if (!image) return null;
    return URL.createObjectURL(image.blob);
  },
};
